### ISSUE TITLE
kubectl edit does not preserve the order of environment variables

### ISSUE THREAD

**Issue by obeattie** [kubernetes/kubernetes#40373]
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):
No

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):
`kubectl edit order`

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
Bug report 🐛 

**Kubernetes version** (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2", GitCommit:"08e099554f3c31f6e6f07b448ab3ed78d0520507", GitTreeState:"clean", BuildDate:"2017-01-12T07:30:54Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.6+coreos.0", GitCommit:"ec2b52fabadf824a42b66b6729fe4cff2c62af8c", GitTreeState:"clean", BuildDate:"2016-11-14T19:42:00Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**:
- **OS** (e.g. from /etc/os-release): Mac OS 10.12.2 (16C67)
- **Kernel** (e.g. `uname -a`): `Darwin oliver.local 16.3.0 Darwin Kernel Version 16.3.0: Thu Nov 17 20:23:58 PST 2016; root:xnu-3789.31.2~1/RELEASE_X86_64 x86_64`
- **Install tools**:
- **Others**:

**What happened**:
When using `kubectl edit` to change the environment variables for a pod, it seems that the _order_ of these variables is not under all conditions preserved. In my case, I was editing the list of environment variables from:

```yaml
- name: SERVICE_PROXY
  value: http://10.1.8.8:80
- name: LISTEN_ADDR
  value: :80
- name: REGION
  value: eu-west-1
```

```yaml
- name: LISTEN_ADDR
  value: :80
- name: REGION
  value: eu-west-1
- name: NODE_NAME
  valueFrom:
    fieldRef:
      apiVersion: v1
      fieldPath: spec.nodeName
- name: SERVICE_PROXY
  value: http://$(NODE_NAME):4140
```

But what I actually ended up with was:

```yaml
- name: SERVICE_PROXY
  value: http://$(NODE_NAME):4140
- name: LISTEN_ADDR
  value: :80
- name: REGION
  value: eu-west-1
- name: NODE_NAME
  valueFrom:
    fieldRef:
      apiVersion: v1
      fieldPath: spec.nodeName
```

Note that the `SERVICE_PROXY` variable appeared *first* rather than *last* in the resulting list. When using environment variable expansion as I am, this is significant: in order for expansion to work correctly, `NODE_NAME` must come before the definition for `SERVICE_PROXY`.

I suspect this is not limited to environment variables and probably applies to any kind of list resource. I haven't had time to dig into the code yet but I suspect that the way the patch is being generated is incorrect.

**What you expected to happen**:
The order of environment variables is as described.

**How to reproduce it** (as minimally and precisely as possible):
1. Create a deployment whose pods includes two environment variables (let's say `A` and `B`)
2. Edit the deployment to reverse the order to `B`, `A`
3. Observe that the saved order is still `A`, `B`

If need be, I can post a more concrete test case.

**Comment by smarterclayton** [comment]
I was able to reproduce this.  Setting priority high because this can actually end up breaking apps (order of env var is part of the API spec).

**Comment by ymqytw** [comment]
@obeattie We don't preserve the order of the environment variables, currently we sort them by the name which is the `mergeKey`.

Here we have at least 2 options to solve this issue:

**1. Always send the entire list**:

We always send the whole list from local file config in the patch. For the entries that have no change, we  only include the `mergeKey` in the list. For the entries that have change, we include the changes as before.

The entries that are not in the patch will be appended after the entries in the patch. The order of the entries that are not in the patch will be preserved.

Example:
Local file config:
```yaml
env:
- name: B
  value: yy
- name: A
  value: xx
- name: D
  value: dd
- name: C
  value: cc
```
Last applied config:
```yaml
env:
- name: A
  value: aa
- name: B
  value: bb
- name: C
  value: cc
```
Live object:
```yaml
env:
- name: A
  value: aa
- name: B
  value: bb
- name: C
  value: cc
- name: F
  value: ff
- name: E
  value: ee
```
Patch:
```yaml
env:
- name: B
  value: yy
- name: A
  value: xx
- name: D
  value: dd
- name: C
```
Result after merging:
```yaml
env:
- name: B
  value: yy
- name: A
  value: xx
- name: D
  value: dd
- name: C
  value: cc
- name: F
  value: ff
- name: E
  value: ee
```
This is backward compatible

**2. Keep the order in the sublist**

Keep the order in the sublist that separated by the items in the patch.
The order in the patch is guaranteed, but the total order has no guarantee.

Example:
Local file config:
```yaml
env:
- name: D
  value: yy
- name: A
  value: xx
```
Last applied config:
```yaml
env:
- name: A
  value: aa
- name: D
  value: dd
```
Live Object:
```yaml
env:
- name: A
  value: aa
- name: B
  value: bb
- name: C
  value: cc
- name: D
  value: dd
- name: E
  value: ee
- name: F
  value: ff
```
Patch:
```yaml
- name: D
  value: yy
- name: A
  value: xx
```
Result after merging:
```yaml
- name: D
  value: yy
- name: E
  value: ee
- name: F
  value: ff
- name: A
  value: xx
- name: B
  value: bb
- name: C
  value: cc
```

This is also backward compatible.

I'm leaning to option 1.

**Comment by ymqytw** [comment]
cc: @pwittrock

**Comment by pwittrock** [comment]
@ymqytw One thing I was thinking, is we can send a parallel list that contains the absolute ordering of all the keys.  If it is present then we use that ordering, and put all items not present in the absolute ordering at the end (maintaining their same relative order).  That maintain backward compatibility while fixing the issue for a new client+server.  That would provide the correct behavior for `kubectl edit` and `kubectl apply`.  We could also do something slightly more sane when the parallel list is missing.

**Comment by ymqytw** [comment]
@pwittrock I think this is better.

**Comment by pwittrock** [comment]
Alright, lets see if we can get this merged by Wednesday.  If not, lets take it out of the milestone.

**Comment by ymqytw** [comment]
SGTM

**Comment by ymqytw** [comment]
@pwittrock Are we going to do this for both list of maps and list of primitives?

**Comment by pwittrock** [comment]
I think that makes sense to do, but don't want to block the env piece on it.

**Comment by pwittrock** [comment]
Given some of the other stuff going into the 1.6 release, this may miss 1.6.

**Comment by ymqytw** [comment]
@smarterclayton I don't think I can make it in 1.6, there is a lot code change, it may be too risky to do this in 1.6.
Is it OK to fix it in 1.7?
